### PR TITLE
Swapstick

### LINKED ISSUE #860: Can't approve token for swaps
Just tried to swap some KEEP to renZEC. Results below

Screenshot from 2022-01-20 18-08-06
Screenshot from 2022-01-20 18-09-37
Screenshot from 2022-01-20 18-10-05

While I couldn't get a screenshot fast enough, it appears the "Swap assets" page flashes at the bottom before the extension blanks. There aren't any obviously relevant errors in the frontend or background console, which surprised me tbh

Tested against `swaps-qa-build` on Chromium 97.0.4692.71 / Ubuntu

### LINKED ISSUE #862: Swaps - Can't swap ETH for ERC20
### Discord Discussion Link

_No response_

### What browsers are you seeing the problem on?

Chrome

### What were you trying to do?

Swap ETH for any ERC20

### What did not work?

When I have ETH selected as my sell currency, the buy dropdown won't populate with tokens 

https://user-images.githubusercontent.com/7005061/150449021-5d569b51-327e-4df0-b890-31a3a1f66ccc.mov

### Version

_No response_

### Relevant log output

_No response_

### LINKED ISSUE #864: Swaps - Unable to edit transaction speed
### Discord Discussion Link

_No response_

### What browsers are you seeing the problem on?

Chrome

### What were you trying to do?

Edit transaction speed / gas cost

### What did not work?

I select different txn speeds, but after clicking "Save," nothing updates on the main Swaps page

Also... the main Swaps page displays a "Fast" option. This isn't one of the options on the txn speed settings page.

https://user-images.githubusercontent.com/7005061/150449361-54851942-c7fc-4aa4-974c-5425a4691e13.mov

### Version

_No response_

### Relevant log output

_No response_

### LINKED ISSUE #866: Swaps - no feedback if user enters amount of sell asset larger than current balance
### Discord Discussion Link

_No response_

### What browsers are you seeing the problem on?

Chrome

### What were you trying to do?

I entered 10000+ KEEP but only have a balance of 470; I expect to get some warning that I don't have this much KEEP. Including either a notification I've put in more than I have and the 'get final quote' button being disabled.

### What did not work?

I get no feedback. I am allowed to attempt the transaction. It loads/spins forver.

https://user-images.githubusercontent.com/7005061/150449927-83bacff0-a53f-4836-8227-b9d21019b09a.mov

### Version

_No response_

### Relevant log output

_No response_

### LINKED ISSUE #868: Swaps - exchange rates are dramatically off & can't complete swap
### Discord Discussion Link

_No response_

### What browsers are you seeing the problem on?

Chrome

### What were you trying to do?

Trade 100 KEEP for USDC or DAI

### What did not work?

- Exchange rate is dramatically off: UI displays 19k USDC whereas 100 KEEP should be ~$60
- This is corrected on approval page...
- ... but after I click "Approve" page goes blank and Swap never completes and never appears in my activity

https://user-images.githubusercontent.com/7005061/150450689-9ab4f467-18ed-4013-856d-43d9a07f2080.mov

### Version

_No response_

### Relevant log output

_No response_

### LINKED ISSUE #869: Swaps - middle 'flip' button does not work
### Discord Discussion Link

_No response_

### What browsers are you seeing the problem on?

Chrome

### What were you trying to do?

Flip the position of buy asset and sell asset

### What did not work?

Button does not appear to have any impact (this could be b/c there is no market for KEEP OXT?)

See 30 secs in video below (I have 27 OXT and a few hundred KEEP):

https://user-images.githubusercontent.com/7005061/150688466-f6036eae-7fed-484b-bef2-da74532bf454.mov

### Version

_No response_

### Relevant log output

_No response_

### LINKED ISSUE #871: Swaps - Not able to get to approval page w/OXT
### Discord Discussion Link

_No response_

### What browsers are you seeing the problem on?

Chrome

### What were you trying to do?

Swap OXT

### What did not work?

I enter 25 OXT (I have 27) as my sell asset, but the buy asset amount remains zero

https://user-images.githubusercontent.com/7005061/150452579-ac255810-c398-49c2-8283-dc9275a2144a.mov

### Version

_No response_

### Relevant log output

_No response_

### LINKED ISSUE #876: Swap - Trade fails on spinner
### Discord Discussion Link

_No response_

### What browsers are you seeing the problem on?

Chrome

### What were you trying to do?

Swap OXT for CRV

### What did not work?

On Sign Transaction page, Spend Amount is 0, when I click "Sign", the loading spinner spins forever and txn does not complete 

https://user-images.githubusercontent.com/7005061/150688861-72f633b8-4e4d-4427-8599-7c37b9ef7a78.mov

 

### Version

_No response_

### Relevant log output

_No response_

### LINKED ISSUE #877: Swap - trade fails on empty activity page
### Discord Discussion Link

_No response_

### What browsers are you seeing the problem on?

Chrome

### What were you trying to do?

Swap KEEP for BAL

### What did not work?

On Sign Transaction page, Amount is Zero. After I click Sign, I'm taken to an empty activity page. Trade does not execute.

https://user-images.githubusercontent.com/7005061/150688986-c5d589d9-cb6b-4b80-bba5-d892b5e9c3fc.mov

### Version

_No response_

### Relevant log output

_No response_

### LINKED ISSUE #928: Final Quote not loading in swaps
### Discord Discussion Link

https://discord.com/channels/808358975287722045/888500174685614090/917871963077218324

### What browsers are you seeing the problem on?

Chrome

### What were you trying to do?

When I am trying to swap my KEEP tokens for another tokem (no matter which one)
nothing pops up for the final quote

### What did not work?

nothing pops up for the final quote

### Version

Community Edition v0.11.0

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by Shadowfiend**
This PR wraps up some of the core work for swaps. The high-level
changes:

- RFQ-T is now fully used throughout our request cycle with 0x. The
  way this worked was a bit different than what was implemented in
  last week's work, as RFQ-T indicative quotes require using a different
  endpoint (`/price`) with a different response schema.
- Swap transaction settings are now fully plugged in and working,
  except for slippage tolerance which is fixed at 1%.
- Swaps can be flipped using the flip icon between sell and buy
  assets.

Relatedly, Chrome bigint support in async thunk responses is now
fixed by patching webext-redux to use our serializers and deserializers
when handling direct responses to dispatched actions.

A few things that are still missing after this:

- Slippage tolerance editing.
- Better loading indication during quoting.
- Some sort of spinner when waiting on an approval transaction to clear.
- Snackbar message when the swap transaction goes out + navigate to
  transaction list.

----

### Testing

To test, run a few quotes and change both the sell and buy assets and
amounts and make sure quotes are showing up. Then try to get a final
quote and swap, or if needed to do a token approval and then get a
final quote and swap. Try both with cases where the sell token must be
approve and with cases where it is already approved to make sure both
flows work as expected.

Closes #877, closes #876, closes #871, closes #869, closes #868, closes #866, closes #864, closes #862, closes #860.

Closes #928.

### CODE REVIEW

**Comment by greg-nagy** [comment]
The good news is that the `webext-patch` works like a charm!  I was about to ask sadly what I am doing wrong when it hit me, that I need to `yarn install` for the patch to run. 

This is what I saw without it: https://cln.sh/jA0Zvq

Getting into testing :) Exciting times

**Comment by Shadowfiend** [comment]
Yah that's ol' #928 😬

File: background/main.ts [hunk 1]
```diff
@@ -365,6 +365,26 @@ export default class Main extends BaseService<never> {
     wrapStore(this.store, {
       serializer: encodeJSON,
       deserializer: decodeJSON,
+      dispatchResponder: async (
```
Comment (greg-nagy) [inline background/main.ts:368]: When is this used? What are those edge cases?

File: patches/webext-redux+2.1.7.patch [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+     });
++    this.deserializer = deserializer
+     this.serializedMessageSender = (0, _serialization.withSerializer)(serializer)(function () {
+       var _this$browserAPI$runt;
+ 
+@@ -231,6 +232,13 @@ function () {
+           var error = resp.error,
+               value = resp.value;
+ 
++          if (typeof _this3.deserializer !== "undefined") {
```
Comment (greg-nagy) [inline patches/webext-redux+2.1.7.patch:17]: It definetly feels like this shouldn't happen :)

File: ui/components/Swap/SwapTransactionSettingsChooser.tsx [hunk 3]
```diff
@@ -1,40 +1,57 @@
...
+  const [networkSettings, setNetworkSettings] = useState(
+    useBackgroundSelector(selectDefaultNetworkFeeSettings)
+  )
 
   const [isSlideUpMenuOpen, setIsSlideUpMenuOpen] = useState(false)
-  const [gasLimit, setGasLimit] = useState("")
-  const dispatch = useDispatch()
 
-  function openSettings() {
+  const openSettings = () => {
```
Comment (greg-nagy) [inline ui/components/Swap/SwapTransactionSettingsChooser.tsx:40]: Was this just a stylistic change or it was required for some reason?

**Review by greg-nagy — COMMENTED** [review]
⛴   ⇋ 🛳  ⇋ 🚢  ⇋ 🛸

**Review by greg-nagy — APPROVED** [review]
⛴ ⇋ 🛳 ⇋ 🚢 ⇋ 🛸

wrong button :(

**Pull request merged**